packages/components/src/components/Combobox/Combobox.tsx

```diff
@@ -0,0 +1,148 @@
+    placeholder?: string;
+    /** The message to display when there are no results. */
+    noResultsMessage?: string;
+    /**
+     * The function to filter the items based on the input value.
+     * If it is true, it will filter the items based on the input value.
+     * If it is a function, it will filter the items based on the input value and the item.
+     */
+    filterItem?: boolean | ((inputValue: string, item: T) => boolean);
+    /** Indicates if the combobox has an error. */
+    hasError?: boolean;
+    /** A custom store to be used instead of the default one. */
+    customStore?: AriakitComboboxProps["store"];
+  };
+
+const matchSorter = (
+  list: ComboboxItemProps[],
+  value: string,
+): ComboboxItemProps[] => {
+  return filter(list, (item) => includes(toLower(item.label), toLower(value)));
+};
+
+const getFilteredItems = <T extends ComboboxItemProps>(
+  items: T[],
+  searchValue: string,
+  filterItem?: boolean | ((inputValue: string, item: T) => boolean),
+): T[] => {
+  if (isBoolean(filterItem)) {
+    return filterItem
+      ? (matchSorter(items as ComboboxItemProps[], searchValue) as T[])
+      : items;
+  }
+  if (isFunction(filterItem)) {
+    return filter(items, (item) => filterItem(searchValue, item));
+  }
+  return items;
+};
+
+// eslint-disable-next-line jsdoc/require-param
+/**
+ * Combobox combines a text input and a popover to allow users to filter from long lists of options.
+ */
+const ComboboxInner = <T extends ComboboxItemProps>(
+  {
+    items,
+    onChangeInput,
+    onSelectItem,
+    placeholder = "Select an option",
+    noResultsMessage = "No results found",
+    filterItem = true,
+    hasError,
+    customStore,
+    ...props
+  }: ComboboxProps<T>,
+  ref: React.ForwardedRef<HTMLInputElement>,
+): React.JSX.Element => {
+  const store = useComboboxStore({
+    store: customStore,
+  });
+  const selectedValue = store.getState().selectedValue;
+  const [searchValue, setSearchValue] = useState("");
+  const [selectedLabel, setSelectedLabel] = useState<string | undefined>("");
+
+  const filteredItems = getFilteredItems(items, searchValue, filterItem);
+
+  return (
+    <ComboboxProvider
+      setSelectedValue={(value) => onSelectItem?.(value)}
+      setValue={(value) => {
+        const label = find(items, ["value", value])?.label;
+        setSearchValue(value);
+        setSelectedLabel(label);
+      }}
+      store={store}
+      value={selectedLabel}
+    >
+      <ComboboxInput
+        aria-label="Select an option"
+        hasError={hasError}
+        onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
+          onChangeInput?.(event);
+          setSelectedLabel(event.target.value);
+        }}
+        placeholder={placeholder}
+        ref={ref}
+        store={store}
+        textOverflow="ellipsis"
+        {...props}
+      />
+      <ComboboxPopover sameWidth>
+        {filteredItems.length === 0 ? (
+          <ComboboxItem>{noResultsMessage}</ComboboxItem>
```

Comment:
isn't it a little bit confusing that we show this message but we still allow users to enter values that are not on the list?

Reply:
Why do you think it's confusing? What do you think it should happen? 

For me, it's clearer to show a message when there are no results for an entry. I think it opens the possibility to add new values when we don't have them. Also, we already have this behavior in our current "Autocomplete" component and I wanted to keep it. Besides that, this is only for the `Combobox` component. The consumer is still able to compose a new component and avoid using the no results message.